salt/utils/dns.py

```diff
@@ -1,18 +1,787 @@
+HAS_NSLOOKUP = salt.utils.which('nslookup') is not None
 
+__salt__ = {
+    'cmd.run_all': salt.modules.cmdmod.run_all
+}
 log = logging.getLogger(__name__)
+ppr = pprint.PrettyPrinter(indent=2).pprint
+
+
+class RFC(object):
+    # http://www.iana.org/assignments/dns-sshfp-rr-parameters/dns-sshfp-rr-parameters.xhtml
+    SSHFP_ALGO = OrderedDict((
+        (1, 'rsa'),
+        (2, 'dsa'),
+        (3, 'ecdsa'),
+        (4, 'ed25519'),
+    ))
+
+    SSHFP_HASH = OrderedDict((
+        (1, 'sha1'),
+        (2, 'sha256'),
+    ))
+
+    # http://www.iana.org/assignments/dane-parameters/dane-parameters.xhtml
+    TLSA_USAGE = OrderedDict((
+        (0, 'pkixta'),
+        (1, 'pkixee'),
+        (2, 'daneta'),
+        (3, 'daneee'),
+    ))
+
+    TLSA_SELECT = OrderedDict((
+        (0, 'pkixta'),
+        (1, 'pkixee'),
+        (2, 'daneta'),
+        (3, 'daneee'),
+    ))
+
+    TLSA_MATCHING = OrderedDict((
+        (0, 'full'),
+        (1, 'sha256'),
+        (2, 'sha512'),
+    ))
+
+    SRV_PROTO = (
+        'tcp',
+        'udp',
+        'sctp'
+    )
+
+    @staticmethod
+    def validate(lookup, ref, match=None):
+        if lookup in ref:
+            return lookup
+        elif match == 'in':
+            return [code for code, name in ref.items() if lookup in name][-1]
+        else:
+            # OrderedDicts only!(?)
+            return ref.keys()[ref.values().index(lookup)]
+
+
+def _data2rec(rschema, rdata):
+    '''
+    OrderedDict({
+        'prio': int,
+        'weight': int,
+        'port': to_port,
+        'name': str,
+    })
+    '10 20 25 myawesome.nl'
+
+    '''
+    try:
+        rdata = rdata.split(' ', len(rschema))
+        rschema = rschema.items()
+        return dict((
+            (fname, rcb(rdata)) for (fname, rcb), rdata in zip(rschema, rdata)
+        ))
+    except:  # pylint: disable=bare-except
+        log.error('Cant parse DNS record data: {}'.format(rdata))
+        return False
+
+
+def _rec2data(*rdata):
+    return ' '.join(rdata)
+
+
+def _to_port(port):
+    try:
+        port = int(port)
+        assert 1 <= port <= 65535
+        return port
+    except (ValueError, AssertionError):
+        raise ValueError('Invalid port {0}'.format(port))
+
+
+def _query_simple(name, rdtype, timeout=None):
+    '''
+    Use Python's socket interface to lookup addresses
+    :param name: Name of record to search
+    :param rdtype: A or AAAA
+    :param timeout: ignored
+    :return: [] of addresses or False if error
+    '''
+    sock_t = {
+        'A': socket.AF_INET,
+        'AAAA': socket.AF_INET6
+    }[rdtype]
+
+    if timeout:
+        log.warn('Ignoring timeout on simple resolver; fix resolv.conf to do that')
+
+    try:
+        addresses = [sock[4][0] for sock in socket.getaddrinfo(name, None, sock_t, 0, socket.SOCK_RAW)]
+        return addresses
+    except socket.gaierror:
+        return False
+
+
+def _query_dig(name, rdtype, timeout=None, servers=[], secure=None):
```

Comment:
We should not have mutable types as default values for arguments. We should change this default value to ``None`` and then do a check at the beginning of the function like this:

```python
if servers is None:
    servers = []
```